File: sandbox/task_pid_positional.c
#include <stdio.h>
#include <math.h>

typedef struct {
    float Kp;
    float Ki;
    float Kd;
    float epsilon;
    float last_error;
    float integral;
    float last_measurement;
} PID_Controller;

void PID_Init(PID_Controller *pid, float Kp, float Ki, float Kd, float epsilon) {
    pid->Kp = Kp;
    pid->Ki = Ki;
    pid->Kd = Kd;
    pid->epsilon = epsilon;
    pid->last_error = 0.0f;
    pid->integral = 0.0f;
    pid->last_measurement = 0.0f;
}

float PID_Calculate(PID_Controller *pid, float setpoint, float measurement) {
    float error = setpoint - measurement;
    
    if (fabsf(error) > pid->epsilon) {
        pid->integral = 0.0f;
    } else {
        pid->integral += error;
    }
    
    float derivative = measurement - pid->last_measurement;
    
    float output = pid->Kp * error + pid->Ki * pid->integral - pid->Kd * derivative;
    
    pid->last_error = error;
    pid->last_measurement = measurement;
    
    return output;
}

int main() {
    PID_Controller pid;
    PID_Init(&pid, 1.2f, 0.5f, 0.3f, 10.0f);
    
    float test_setpoints[] = {100.0f, 100.0f, 100.0f, 100.0f, 100.0f};
    float test_measurements[] = {80.0f, 85.0f, 95.0f, 99.0f, 100.5f};
    
    printf("周期号 | 设定值 | 测量值 | 控制输出\n");
    printf("------------------------------------\n");
    
    for (int i = 0; i < 5; i++) {
        float output = PID_Calculate(&pid, test_setpoints[i], test_measurements[i]);
        printf("%6d | %6.1f | %6.1f | %8.2f\n", 
               i + 1, test_setpoints[i], test_measurements[i], output);
    }
    
    return 0;
}

File: sandbox/task_pid_positional.py
import math

class PID_Controller:
    def __init__(self, Kp, Ki, Kd, epsilon):
        self.Kp = Kp
        self.Ki = Ki
        self.Kd = Kd
        self.epsilon = epsilon
        self.last_error = 0.0
        self.integral = 0.0
        self.last_measurement = 0.0

def PID_Calculate(pid, setpoint, measurement):
    error = setpoint - measurement
    
    if abs(error) > pid.epsilon:
        pid.integral = 0.0
    else:
        pid.integral += error
    
    derivative = measurement - pid.last_measurement
    
    output = pid.Kp * error + pid.Ki * pid.integral - pid.Kd * derivative
    
    pid.last_error = error
    pid.last_measurement = measurement
    
    return output

def main():
    pid = PID_Controller(1.2, 0.5, 0.3, 10.0)
    
    test_setpoints = [100.0, 100.0, 100.0, 100.0, 100.0]
    test_measurements = [80.0, 85.0, 95.0, 99.0, 100.5]
    
    print("周期号 | 设定值 | 测量值 | 控制输出")
    print("------------------------------------")
    
    for i in range(5):
        output = PID_Calculate(pid, test_setpoints[i], test_measurements[i])
        print(f"{i+1:6d} | {test_setpoints[i]:6.1f} | {test_measurements[i]:6.1f} | {output:8.2f}")

if __name__ == "__main__":
    main()